Cookbook/Proposals/SwiftGen.md

```diff
@@ -0,0 +1,83 @@
+## Adopt SwiftGen for localizable strings
+
+* Author: David Rodrigues
+* Review Manager: Danilo Aliberti
+
+### Introduction
+
+Localisable strings by themselves are not a hard thing to use but for apps with a significant number of strings it's entirely impossible to remember or discover easily the respective key to use.
+
+```swift
+title = NSLocalizedString("hello_world_title", comment: "")
+```
+
+Because of this, and considering the size of our codebase, we use concrete types that map our localisable strings which enables the usage of autocomplete.
+
+```swift
+public struct LocalizationUI {
+    public struct Common {
+        public static var ok: String { return localizedName(key: "general_ok") }
+        ...
+    }
+}
+```
+
+### Problems
+
+The current approach works, as we would expect, but there's a couple of caveats in it.
+
+##### 1. Maintenance
+
+This types at the moment are created and managed by us manually which means any new addition/edition/removal needs to be done by us. This may not look very significant but considering the number of strings, more than 3k, it is a lot to be maintained/extended.
+
+##### 2. Safety
+
+While this types are useful for development by enabling autocomplete they don't provide any safety since we can add a new property for a new key without necessarily adding the new string.
+
+### Proposed Solution
+
+SwiftGen is a code generator which is able to generate all this types and properties automatically by running a script which would fix the two main problems that we have with the current approach.
+
+- Maintenance will be zero because everything will be mapped based on the localisable strings;
+- Compile-time type safety will be guaranteed because everything will be generated from the localizable strings making it impossible to use something without being in the strings file or keep using something that was deleted.
+
+SwiftGen also supports strings with dynamic input generating a function to be called with all the parameters which also ensures safety as we clearly know that a certain string requires some dynamic values.
+
+We still need to ensure this script is correctly executed every time that we add/edit/remove a string but that's a easier task to manage over the current approach.
+
+### Implementation
+
+Adopting SwiftGen won't be a trivial task and would require a certain level of investment from us.
+
+##### Namespacing
+
+Based on how we have our keys defined we won't be able to mangle them into namespaces directly, we will have a huge flat list with all the strings. This can be changed but requires a major refactor of the keys to define the respective vertical/domain/module.
+
+As an example, `chatbot_send` should become `chatbot.send` to ensure we get a namespace for chatbot. Ideally we would like to keep the same namespaces to avoid a huge refactor in the codebase.
```

Comment:
I was thinking about the same issue, but can't we instead adjusts the templates to treat `_`, let's say just first one to avoid "over namespacing" as `.`?

Reply:
The split against `.` is done inside the SwiftGen engine (and then the templates can use it to loop over each groups to generate the code).

Maybe we can make the template do the split only for first level, but the split won't be easy to code via a template in a recursive way. There's some pending work to allow customization of the engine itself (similar to `params:` in the config file, but for arguments passed to the parser engine, not to the template) in order, typically for Strings, to provide a different spilt character (`_` instead of `.`) that would solve that. That require that I find some time to triage some pending PRs and make the overdue release and implement this feature first though (I really need to find some time to work on SwiftGen again, to review pending PRs and releases...)

Reply:
|If the only option then is to update the keys then we can do that with an API call https://lokalise.co/api2docs/curl/#transition-update-a-key-put

Reply:
1) list all keys
https://lokalise.co/api2docs/curl/#transition-list-all-keys-get
2) bulk update
https://lokalise.co/api2docs/curl/#transition-bulk-update-put
3) run to the hills 🏃🤣

Reply:
Updating the keys will probably mean breaking Android too. And in Android, dots in keys are problematic (since `R.java` doesn't expect dots in key names)

Reply:
I don't think we are using the same keys between Android and iOS, at least they were always defined without any synchronisation AFAIK, @daniloaliberti can you confirm?

Reply:
Android and iOS have different projects in Lokalise, so I don't see that there would be a problem.

Reply:
Oh. you mean each text in the app appears twice (and thus is potentially translated twice) in Lokialize, once for the Android screen and once for the iOS screen, even though they are likely to be the same translations?

If so, that's great for us (because it would ease the transition to SwiftGen), but not sure how translators are coping with that double-work and handling match translations in common iOS/Android screens?

Reply:
BTW I remember now that there are some limitations with Lokalize and how they treat placeholders as opposed to literal percent characters. Especially, iirc, in Lokalize to specify a placeholder, you'd wrap it inside square brackets to tell it's a placeholder, while literal percents would be `[%]` – see https://docs.lokalise.co/en/articles/1400511-lokalise-placeholders

So you could have:

String in Lokalize | String exported in `Localizable.strings` | Intended call site | What SwiftGen sees
---|---|---|---
`"Hello [%@]!"` | `"Hello %@"` | Using `String(format:)` | contains a placeholder, will use `String(format:)`
`"Progress: [%d][%]"` | `"Progress: %d%%"` | Using `String(format:)` | contains placeholders, will use `String(format:)`
`"Yay, 10% discount!"` | `"Yay, 10% discount!"` | Using directly, no format | Contains a `% d`, which is a valid placeholder, so generates a function taking an `Int` and considers it needs to call `String(format:)` with it

So the problem is that when SwiftGen sees any translation containing a valid `%...` token as per the printf-spec (e.g. `%2d`, `%@`, `%3 d`...) it considers the localisation will need to take parameters and would need to go thru `String(format:)`, so it will generate a `func` taking those parameters. And if you want a literal `%` in there, you're supposed to then use `%%` to "escape" the literal percent character so that it produces a literal `%` after `String(format:)`.

But the problem is that it's not how Lokalise sees things, as **Lokalize expect for you to know at call site when you're supposed to call `String(format:)` for some keys and not use it for others** (see column 2), **based on human knowledge only**. So this decision of calling `String(format:)` or not on each key can't really be automated based on that, there's no real way for SwiftGen to know, when it sees `"Yay, 10% discount!"`, that it's supposed to be a literal and not to be used with `String(format:)`...

What we usually recommend is that every time you have a literal `%` in one of your translation, you should always double it in your `Localizable.strings` as `%%`, so that it's properly considered as a literal by SwiftGen and does not risk to accidentally be considered as a placeholder depending on the character that follow. But that's not how Lokalise see and export things :/

---

So if anyone have any idea how to solve that – either what algorithm we could use to determine **with certainty** if a localisation requires `String(format:)` or not based on what Lokalise exports, or a way to provide this info as some kind of metadata, or a way to hijack / post-process `lokalise pull` to replace every string that's considered to not require `String(format:)` by Lokalise but still contains `%`, to transform them to use `%%` instead... maybe with the Lokalise API to determine for each key if it exposes that it contains `[%...]` or not before export into `.strings` files...? – in any way I'm interested in any idea about this.

For more info on this limitation, see https://github.com/SwiftGen/SwiftGen/issues/577#issuecomment-455270805

Reply:
Mmmh after digging Lokalise might have fixed that since last time I looked at the issue (see my last comment on the SwiftGen issue), so maybe it's a false alarm after all 😅 but we definitively need to keep that case in mind and test it.

Reply:
> Updating the keys will probably mean breaking Android too. And in Android, dots in keys are problematic (since R.java doesn't expect dots in key names)

> Android and iOS have different projects in Lokalise, so I don't see that there would be a problem.
  
It's possible to define different keys by platform, so it won't be an issue even if we use the same project.